metals/src/main/scala/scala/meta/internal/metals/debug/DebugDiscovery.scala

```diff
@@ -162,9 +172,278 @@ class DebugDiscovery(
+        createTestCaseParams(buildTarget, suiteFqn, testName)
+      case None =>
+        Future.failed(NoRunOptionException)
+    }
+  }
+
+  private def findMainClasses(
+      textDocument: TextDocument,
+      classes: BuildTargetClasses.Classes,
+  ): Seq[b.ScalaMainClass] = {
+    val mains = for {
+      occurrence <- textDocument.occurrences
+      if occurrence.role.isDefinition || occurrence.symbol == "scala/main#"
+      symbol = occurrence.symbol
+      mainClass <- {
+        val normal = classes.mainClasses.get(symbol)
+        val fromAnnot = DebugDiscovery
+          .mainFromAnnotation(occurrence, textDocument)
+          .flatMap(classes.mainClasses.get(_))
+        List(normal, fromAnnot).flatten
+      }
+    } yield mainClass
+
+    if (mains.nonEmpty) mains
+    else DebugDiscovery.syntheticMains(textDocument, classes)
+  }
+
+  private def discoverTestCases(
+      path: AbsolutePath
+  ): List[(String, List[TestCaseEntry])] = {
+    import scala.jdk.CollectionConverters._
+    import scala.meta.internal.metals.testProvider.TestExplorerEvent.AddTestCases
+
+    val testUpdates = testProvider.discoverTests(Some(path))
+    testUpdates
+      .flatMap(_.events.asScala)
+      .collect { case AddTestCases(fqn, _, testCases) =>
+        (fqn, testCases.asScala.toList)
+      }
+      .toList
+  }
+
+  private def calculateMainDistances(
+      textDocument: TextDocument,
+      mains: Seq[b.ScalaMainClass],
+      classes: BuildTargetClasses.Classes,
+      position: Position,
+  ): Seq[Distance] = {
+    mains.flatMap { mainClass =>
+      val occOpt = textDocument.occurrences.find { occ =>
+        occ.role.isDefinition && (
+          classes.mainClasses.get(occ.symbol).contains(mainClass) ||
+            DebugDiscovery
+              .mainFromAnnotation(occ, textDocument)
+              .flatMap(classes.mainClasses.get(_))
+              .contains(mainClass)
+        )
+      }
+
+      val syntheticOccOpt = if (occOpt.isEmpty) {
+        val mainMethodName = mainClass.getClassName.split('.').last
+
+        val scala3MainOpt = textDocument.symbols
+          .find { symbolInfo =>
+            symbolInfo.symbol.contains(s".$mainMethodName().") &&
+            symbolInfo.annotations.exists { annotation =>
+              annotation.tpe match {
+                case tpe: scala.meta.internal.semanticdb.TypeRef =>
+                  tpe.symbol == "scala/main#"
+                case _ => false
+              }
+            }
+          }
+          .flatMap(symbolInfo =>
+            textDocument.occurrences.find(_.symbol == symbolInfo.symbol)
+          )
+
+        scala3MainOpt.orElse {
+          val expectedSymbol = mainClass.getClassName.replace(".", "/")
+          textDocument.symbols
+            .find { symbolInfo =>
+              symbolInfo.symbol.startsWith(expectedSymbol) &&
+              (symbolInfo.symbol
+                .endsWith("#") || symbolInfo.symbol.endsWith("."))
+            }
```

Comment:
```suggestion
            .find { symbolInfo =>
              symbolInfo.symbol == expectedSymbol + "#" || symbolInfo.symbol == expectedSymbol + "." 
            }
```
wouldn't be more exact?